### PR TITLE
Add init process for zombie fighting and signal handling 

### LINKED ISSUE #3793: docker client not passing signals to dockerd
We'd like a signal to the docker client to trigger a stop, so if the docker client is killed, it cleans up the containers. 

Filing bug per @crosbymichael request for how to reproduce. 

```
$ docker run -sig-proxy=true busybox sleep 3000
 
(other terminal)
$ docker ps
e8767eed1283        busybox:latest      sleep 3000          28 seconds ago      Up 28 seconds                           insane_pasteur      
$ ps aux|grep 'busybox sleep'
  782 docker   docker run -sig-proxy=true  busybox sleep 3000
$ kill 782
$ docker ps
e8767eed1283        busybox:latest      sleep 3000          45 seconds ago      Up 45 seconds                           insane_pasteur   

... still running ...

$ kill -9 782
$ ps aux | grep 782 
$

... docker run (client) has been killed ... but the container is still running.   

$ docker ps
e8767eed1283        busybox:latest      sleep 3000          3 minutes ago       Up 3 minutes                            insane_pasteur     
```  

#1249 appears to have the functionality needed to fix this, but that appears to have been removed. Is there any other way to get this functionality?

### LINKED ISSUE #24284: Executing child process in scripts hangs terminal
**Output of `docker version`:**

```
Client:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   906eacd
 Built:        Fri Jun 17 20:35:33 2016
 OS/Arch:      darwin/amd64
 Experimental: true

Server:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   a7119de
 Built:        Wed Jun 29 10:03:33 2016
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 2
 Running: 1
 Paused: 0
 Stopped: 1
Images: 482
Server Version: 1.12.0-rc2
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 380
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host null overlay
Swarm: inactive
Runtimes: default
Default Runtime: default
Security Options: seccomp
Kernel Version: 4.4.14-moby
Operating System: Alpine Linux v3.4
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.954 GiB
Name: moby
ID: Z25L:V4GW:6DNP:UQGU:2DRW:BROW:BPRA:K3S5:LBE7:PYWL:KOYC:UJ2N
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 112
 Goroutines: 211
 System Time: 2016-07-03T15:47:20.113835804Z
 EventsListeners: 49
No Proxy: *.local, 169.254/16
Username: mckn
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
I am running on OSX 10.11.5 

**Steps to reproduce the issue:**
1. docker run -d --name cassandra cassandra:2
2. wait until started (takes about 10s)
3. docker run -e "DATABASE=cassandra" --link cassandra:kong-database -ti mashape/kong /usr/local/bin/kong start
4. It will hang on [INFO] Auto-generating the default SSL certificate and key...
5. source to that file is: https://github.com/Mashape/kong/blob/997561b42c27d5917f261065d9b1f5e99079e337/kong/cli/utils/ssl.lua#L43
6. It will continue if you press ctrl-c

**Describe the results you received:**
Docker seems to hang when spawning child processes inside executed scripts.

**Describe the results you expected:**
It should continue in the flow and print [OK] Started and then exit the docker run.

**Additional information you deem important (e.g. issue happens only occasionally):**

### PR DESCRIPTION
**Pull request by crosbymichael**
This adds a small C binary for fighting zombies.  It is mounted under
`/dev/init` and is prepended to the args specified by the user.  You
enable it via a daemon flag, `dockerd --init`, as it is disable by
default for backwards compat.

You can also override the daemon option or specify this on a per
container basis with `docker run --init=true|false`.

You can test this by running a process like this as the pid 1 in a
container and see the extra zombie that appears in the container as it
is running.

``` c

int main(int argc, char ** argv) {
    pid_t pid = fork();
    if (pid == 0) {
        pid = fork();
        if (pid == 0) {
            exit(0);
        }
        sleep(3);
        exit(0);
    }
    printf("got pid %d and exited\n", pid);
    sleep(20);
}
```

Fixes #3793
Fixes #24284

Signed-off-by: Michael Crosby crosbymichael@gmail.com

### CODE REVIEW

**Comment by tonistiigi** [comment]
```
# dockerd -s overlay -D --reaper 2>/daemon.log &
# docker run busybox /nosuchfile
error: No such file or directory
# echo $?
1

# dockerd -s overlay -D 2>/daemon.log &
# docker run busybox /nosuchfile
docker: Error response from daemon: oci runtime error: exec: "/nosuchfile": stat /nosuchfile: no such file or directory.
# echo $?
127
```

We should add an option to run integration-cli with both modes. I'm sure we have a test for this case.

**Comment by cpuguy83** [comment]
Looks like this will resolve #3793

File: daemon/oci_linux.go [hunk 1]
```diff
@@ -593,6 +594,19 @@ func (daemon *Daemon) populateCommonSpec(s *specs.Spec, c *container.Container)
...
+		s.Process.Args = append([]string{"/dev/init", c.Path}, c.Args...)
+		path, err := exec.LookPath("docker-init")
+		if err != nil {
+			return err
+		}
+		s.Mounts = append(s.Mounts, specs.Mount{
+			Destination: "/dev/init",
+			Type:        "bind",
+			Source:      path,
+			Options:     []string{"bind", "rw"},
```
Comment (AkihiroSuda) [inline daemon/oci_linux.go]: why rw?

**Comment by crosbymichael** [comment]
Ok, so as far as flags go, @tonistiigi and I talked about this some.  We are not totally sold on what the flag name should be but we have a good idea on how it should work.

On the daemon we will have a bool `--init` flag.  This will be false by default, and if set to true, will run the custom docker init in every container.

On run, we will also have a `--init` flag that is a nullable bool.  If it is not set, we use the daemon default.  If it is set to true, we use the docker init for the container, if it is `--init=false` we will not use the init.  This allows us to have bot a daemon wide option and the ability to enable/disable per container, without changing anything for users that don't have this set.

If anyone has a better idea for a flag name let me know.

**Comment by crosbymichael** [comment]
Updated this to use `--init` for now on both the daemon and run.  You can try out all the various combinations.

**Comment by icecrime** [comment]
Moving this to code review: I think there no big controversy on the design :+1: I agree with everything in that comment https://github.com/docker/docker/pull/26061#issuecomment-243883226.

File: integration-cli/docker_cli_run_test.go [hunk 2]
```diff
@@ -2402,30 +2402,6 @@ func (s *DockerSuite) TestRunExposePort(c *check.C) {
 	c.Assert(out, checker.Contains, "invalid range format for --expose")
 }
 
-func (s *DockerSuite) TestRunUnknownCommand(c *check.C) {
-	out, _, _ := dockerCmdWithStdoutStderr(c, "create", "busybox", "/bin/nada")
```
Comment (tonistiigi) [inline integration-cli/docker_cli_run_test.go]: What if we do a lookup and mode check for the entrypoint before actually executing to maintain this behavior?

Comment (crosbymichael) [inline integration-cli/docker_cli_run_test.go]: Like in the docker daemon?

Comment (tonistiigi) [inline integration-cli/docker_cli_run_test.go]: in daemon or containerd

File: integration-cli/docker_cli_events_test.go [hunk 3]
```diff
@@ -75,40 +75,6 @@ func (s *DockerSuite) TestEventsUntag(c *check.C) {
 	}
 }
 
-func (s *DockerSuite) TestEventsContainerFailStartDie(c *check.C) {
```
Comment (icecrime) [inline integration-cli/docker_cli_events_test.go]: Why is this change related to the PR?

File: integration-cli/docker_cli_run_test.go [hunk 4]
```diff
@@ -2402,30 +2402,6 @@ func (s *DockerSuite) TestRunExposePort(c *check.C) {
 	c.Assert(out, checker.Contains, "invalid range format for --expose")
 }
 
-func (s *DockerSuite) TestRunUnknownCommand(c *check.C) {
```
Comment (icecrime) [inline integration-cli/docker_cli_run_test.go]: Why is this change related to the PR?

**Comment by icecrime** [comment]
Small questions regarding tests, otherwise LGTM :+1:

File: integration-cli/docker_cli_run_test.go (hunk 2 above)
Comment (crosbymichael) [inline integration-cli/docker_cli_run_test.go]: won't work because lookups have to happen inside the container's mount ns after all volumes are mounted because PATH and binaries could live inside them

Comment (tonistiigi) [inline integration-cli/docker_cli_run_test.go]: We already have same hack for `cp`.

Comment (crosbymichael) [inline integration-cli/docker_cli_run_test.go]: yes, "hack"

File: integration-cli/docker_cli_events_test.go (hunk 3 above)
Comment (crosbymichael) [inline integration-cli/docker_cli_events_test.go]: If we run the tests with this feature enabled these two will fail because they check that the cmd passed does not exist and it will always exist because we prepend our init.

It will still exit with a 127 but not at start time.  I tried looking through the history to see why these these were added but I couldn't find anything.  

We went over this same discussion in the runc change for 1.11  and I said that if we were to ever solve the zombie problem this is not going to work.  We get the same outcome, the containers fails with 127 exit status but it does not return the error on start.

Comment (icecrime) [inline integration-cli/docker_cli_events_test.go]: I see, thanks :+1:

File: daemon/oci_linux.go [hunk 5]
```diff
@@ -585,6 +586,26 @@ func (daemon *Daemon) populateCommonSpec(s *specs.Spec, c *container.Container)
 		cwd = "/"
 	}
 	s.Process.Args = append([]string{c.Path}, c.Args...)
+
+	// only add the custom init if it is specified and the container is running in its
+	// own private pid namespace.  It does not make sense to add if it is running in the
+	// host namespace or another container's pid namespace were we already have an init
```
Comment (justincormack) [inline daemon/oci_linux.go]: type "were" not "where"

File: api/types/container/host_config.go [hunk 6]
```diff
@@ -321,4 +321,8 @@ type HostConfig struct {
 
 	// Mounts specs used by the container
 	Mounts []mount.Mount `json:",omitempty"`
+
+	// Run a custom init inside the container, null means the use the daemon's configured settings for
```
Comment (mlaventure) [inline api/types/container/host_config.go]: `s/the use the daemon/we use the daemon/`?

**Comment by justincormack** [comment]
found a typo, but LGTM otherwise.

**Review by icecrime — APPROVED** [review]

**Comment by tianon** [comment]
:heart_eyes: any way we'll be able to turn this on by default at the image metadata level at some point? (since the author of a particular software image is likely to be able to tell whether or not a particular bit of software is going to be making zombies and thus need reaping)

I've got several images that use either `tini` or `dumb-init` for this purpose currently that I'd _love_ to replace with some metadata telling Docker that each image needs this new `/dev/init` behavior. :heart:

**Comment by crosbymichael** [comment]
@tianon yes, we will add a dockerfile instruction

**Comment by crosbymichael** [comment]
fixed the typos , merge'n

File: api/types/container/host_config.go [hunk 7]
```diff
@@ -321,6 +321,9 @@ type HostConfig struct {
 
 	// Mounts specs used by the container
 	Mounts []mount.Mount `json:",omitempty"`
+
+	// Run a custom init inside the container, if null, use the daemon's configured settings
+	Init *bool `json:",om        itempty"`
```
Comment (dsheets) [inline api/types/container/host_config.go]: This looks strange.

Comment (thaJeztah) [inline api/types/container/host_config.go]: Yup, indeed; can you open a PR to fix it?

Comment (dsheets) [inline api/types/container/host_config.go]: Now #26736.

**Comment by erikh** [comment]
congrats!

File: daemon/oci_linux.go [hunk 8]
```diff
@@ -585,6 +586,26 @@ func (daemon *Daemon) populateCommonSpec(s *specs.Spec, c *container.Container)
...
 	s.Process.Args = append([]string{c.Path}, c.Args...)
+
+	// only add the custom init if it is specified and the container is running in its
+	// own private pid namespace.  It does not make sense to add if it is running in the
+	// host namespace or another container's pid namespace where we already have an init
+	if c.HostConfig.PidMode.IsPrivate() {
+		if (c.HostConfig.Init != nil && *c.HostConfig.Init) ||
+			(c.HostConfig.Init == nil && daemon.configStore.Init) {
+			s.Process.Args = append([]string{"/dev/init", c.Path}, c.Args...)
+			path, err := exec.LookPath("docker-init")
```
Comment (runcom) [inline daemon/oci_linux.go]: @crosbymichael can we have a flag to configure the path to this binary?

Comment (justincormack) [inline daemon/oci_linux.go]: @runcom we havent done this for any of the other binaries, I think if the dockerd install is in an odd place then you should change the `PATH` in its init script. We have not had any issues with any of the other binaries.

Comment (justincormack) [inline daemon/oci_linux.go]: ah ok I missed https://github.com/docker/docker/pull/26882

Comment (justincormack) [inline daemon/oci_linux.go]: yeah, ok, probably makes sense, as the user wont execute directly.

Comment (crosbymichael) [inline daemon/oci_linux.go]: @runcom can you explain why this would be helpful or needed?

Comment (runcom) [inline daemon/oci_linux.go]: @crosbymichael to allow installation of those binaries outside $PATH. On certain distribution binaries used only by daemons are installed under /usr/libexec for instance which is not in  $PATH. Another thing is that this can enable multiple docker installations on the same host (not daemon running concurrently, just daemon installation)

Comment (crosbymichael) [inline daemon/oci_linux.go]: ok, that makes sense, SGTM

Comment (runcom) [inline daemon/oci_linux.go]: Alright, thanks, I'll work on a patch to have a flag on the daemon for docker-init

**Comment by asottile** [comment]
Curious why a new implementation of a docker init system was chosen over other implementations which have tests, many more users, and likely fewer bugs such as:
- https://github.com/Yelp/dumb-init/
- https://github.com/krallin/tini

**Comment by rogaha** [comment]
+1, thanks for working on it @crosbymichael! That's awesome!

**Comment by glensc** [comment]
is the `/dev` abuse really a good idea?

i recall discussion about `/run` because vendors abused `/dev` for their early boot dirty work:
https://lists.fedoraproject.org/pipermail/devel/2011-March/150031.html

perhaps just use file named `docker-init` (or `.docker-init`) in root filesystem or `/lib` or `/sbin`?

**Comment by ndeloof** [comment]
That's a nice improvement to get docker handle this PID1 issue. But I'm concerned this moves the responsibility to run with `--init` option into end users hands. For sample, the [jenkins docker image](https://github.com/jenkinsci/docker/blob/master/Dockerfile#L67) uses tiny for same purpose, but end user don't have to even know what a zombie process is.

I suggest one could use some metadata in image to let runtime know such init process is required, for sample by adding to Dockerfile: `LABEL io.docker.init='true'`

**Comment by thaJeztah** [comment]
> I suggest one could use some metadata in image to let runtime know such init process is required, for sample by adding to Dockerfile: LABEL io.docker.init='true'

In cases where a zombie reaper is required (in most cases, it should not even be needed), it should be added by the image author. I think this PR should be seen as a "last effort" if the image is not under control of the user.

**Comment by tianon** [comment]
Did the Dockerfile instruction to enable this by default at the image level
we discussed make it in to 1.13? :innocent:
